### PR TITLE
NEW (CodeAnalyzer): @W-12943227@: Adds sample project for writing Java-based PMD custom rules.

### PR DESCRIPTION
**Pull request by jfeingold35**

### CODE REVIEW

File: sample-code/pmd-example-rules/README.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+Salesforce Code Analyzer supports the use of custom PMD rules. If those rules involve any Java code,
+then that Java must be compiled into a JAR before it can be registered with Code Analyzer.
+
+This project is an example of how to write Java-based PMD rules, and use Maven to compile them into a JAR.
+
+## How do I build this project?
+1. Make sure [Maven](https://maven.apache.org/guides/getting-started/maven-in-five-minutes.html) is set up on your machine.
+2. Clone the Code Analyzer project.
+3. From the `sample-code/pmd-example-rules` directory, run `mvn clean package`.
+4. Observe that a new folder was created named `target`, containing (among other things) a file named `pmd-example-rules-1.0-SNAPSHOT.jar`.
```
[the remark this answers is not in the document]
Comment (teresa-allen-sfdc) [reply sample-code/pmd-example-rules/README.md:11]: @jfeingold35 All edits to be made to this file are in this doc - I used Suggesting Mode:

https://docs.google.com/document/d/1YAOmc63r1y95SD8N-8Ty0n13uPDfNR89rV8ByXLEV0M/edit?usp=sharing

File: sample-code/pmd-example-rules/README.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# PMD example rules
+Salesforce Code Analyzer supports the use of custom PMD rules. If those rules involve any Java code,
```
Comment (teresa-allen-sfdc) [inline sample-code/pmd-example-rules/README.md:2]: PMD Example Rules

Salesforce Code Analyzer supports the use of custom PMD rules. If your custom PMD rules include Java code, then you must compile the Java code into a JAR before you register it with Code Analyzer.

This project is an example of how to write Java-based PMD rules, and use Maven to compile them into a JAR.

Comment (teresa-allen-sfdc) [reply sample-code/pmd-example-rules/README.md:2]: @jfeingold35 See edits in the G doc: https://docs.google.com/document/d/1YAOmc63r1y95SD8N-8Ty0n13uPDfNR89rV8ByXLEV0M/edit?usp=sharing

File: sample-code/pmd-example-rules/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (teresa-allen-sfdc) [reply sample-code/pmd-example-rules/README.md:11]: 1. Install [Maven](https://maven.apache.org/guides/getting-started/maven-in-five-minutes.html).
2. Clone the Code Analyzer [add link] project.
3. From the sample-code/pmd-example-rules directory, run mvn clean package.
4. Notice the new folder named target. It contains a file named pmd-example-rules-1.0-SNAPSHOT.jar.
5. From the sample-code/pmd-example-rules directory, run sfdx scanner:rule:add --language apex --path ./target/pmd-example-rules-1.0-SNAPSHOT.jar.
6. Run scanner:rule:list --category CustomRules --engine pmd to confirm the addition of the new rule.
7. Run scanner:run --category CustomRules --engine pmd --target path/to/any/apexfile.cls and view the new rule violation.

File: sample-code/pmd-example-rules/README.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+1. Make sure [Maven](https://maven.apache.org/guides/getting-started/maven-in-five-minutes.html) is set up on your machine.
+2. Clone the Code Analyzer project.
+3. From the `sample-code/pmd-example-rules` directory, run `mvn clean package`.
+4. Observe that a new folder was created named `target`, containing (among other things) a file named `pmd-example-rules-1.0-SNAPSHOT.jar`.
+5. From the `sample-code/pmd-example-rules` directory, run `sfdx scanner:rule:add --language apex --path ./target/pmd-example-rules-1.0-SNAPSHOT.jar`.
+6. Run `scanner:rule:list --category CustomRules --engine pmd` to confirm the addition of the new rule.
+7. Run `scanner:run --category CustomRules --engine pmd --target path/to/any/apexfile.cls` and observe a violation for the new rule.
+
+## What else can I do with this project?
+You can use this project as a template for how to define and build your own custom rules, categories, and rulesets.
```
[the remark this answers is not in the document]
Comment (teresa-allen-sfdc) [reply sample-code/pmd-example-rules/README.md:17]: Use this project as a template for how to define and build your own custom rules, categories, and rulesets.

**Review by teresa-allen-sfdc — COMMENTED** [review]
Review complete - see G doc for the same edits

File: sample-code/pmd-example-rules/pom.xml [hunk 4]
```diff
@@ -0,0 +1,78 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+This is a sample POM file allowing this module to be compiled and bundled into a JAR via maven.
+If you have Maven properly configured on your machine, you can run `mvn clean package` to build
+a JAR containing everything needed to register and run a custom rule.
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0"
```
Comment (ganeshnt) [inline sample-code/pmd-example-rules/pom.xml:7]: I would think that this sample project was locally built with this pom file and the build works.  I just can't verify the semantic validity of the pom file is all am saying

**Review by ganeshnt — APPROVED** [review]

**Pull request merged**